### PR TITLE
Improve the local optimizer at the end of resolve

### LINKED ISSUE #1949: ]up only downgrades packages
As mentioned on Slack, I have a weird situation where `]up` is doing nothing but downgrading packages, though it doesn't seem like there's any incompatibility problems:

```
(ECHOAnalysis) pkg> up
   Updating registry at `~/.julia/registries/General`
   Updating git-repo `https://github.com/JuliaRegistries/General.git`
Updating `~/repos/scratch/ECHOAnalysis.jl/Project.toml`
  [3bd8f0ae] ↓ Microbiome v0.6.4 ⇒ v0.6.1
Updating `~/repos/scratch/ECHOAnalysis.jl/Manifest.toml`
  [b4f34e82] - Distances v0.9.0
  [3bd8f0ae] ↓ Microbiome v0.6.4 ⇒ v0.6.1
  [3cdcf5f2] ↓ RecipesBase v1.0.2 ⇒ v0.8.0
  [348f2d5d] ↓ SpatialEcology v0.9.5 ⇒ v0.7.0
  [2913bbd2] ↓ StatsBase v0.33.0 ⇒ v0.32.2
```

This is just a personal package and the issue is easily solved by adding Microbiome 0.6.4 to the deps, so it's not urgent for me, but seems like an odd bug. I'm wondering if it might be a downstream consequence of the move to the BioJulia registry and then back to General, coupled with something I did here: https://github.com/JuliaRegistries/General/pull/10324

Steps to reproduce:

```
$ git clone git@github.com:Klepac-Ceraj-Lab/ECHOAnalysis.jl.git # current commit 4f3e59b
Cloning into 'ECHOAnalysis.jl'...
remote: Enumerating objects: 233, done.
remote: Counting objects: 100% (233/233), done.
remote: Compressing objects: 100% (118/118), done.
remote: Total 1840 (delta 118), reused 193 (delta 83), pack-reused 1607
Receiving objects: 100% (1840/1840), 12.91 MiB | 7.24 MiB/s, done.
Resolving deltas: 100% (1134/1134), done.
$ cd ECHOAnalysis.jl 
$ julia --project
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.5.0 (2020-08-01)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

(ECHOAnalysis) pkg> add Microbiome@0.6.4
  Resolving package versions...
Updating `~/repos/scratch/ECHOAnalysis.jl/Project.toml`
  [3bd8f0ae] ↑ Microbiome v0.6.1 ⇒ v0.6.4
Updating `~/repos/scratch/ECHOAnalysis.jl/Manifest.toml`
  [b4f34e82] + Distances v0.9.0
  [3bd8f0ae] ↑ Microbiome v0.6.1 ⇒ v0.6.4
  [3cdcf5f2] ↑ RecipesBase v0.8.0 ⇒ v1.0.2
  [348f2d5d] ↑ SpatialEcology v0.7.0 ⇒ v0.9.5
  [2913bbd2] ↑ StatsBase v0.32.2 ⇒ v0.33.0

(ECHOAnalysis) pkg> up
   Updating registry at `~/.julia/registries/General`
   Updating git-repo `https://github.com/JuliaRegistries/General.git`
Updating `~/repos/scratch/ECHOAnalysis.jl/Project.toml`
  [3bd8f0ae] ↓ Microbiome v0.6.4 ⇒ v0.6.1
Updating `~/repos/scratch/ECHOAnalysis.jl/Manifest.toml`
  [b4f34e82] - Distances v0.9.0
  [3bd8f0ae] ↓ Microbiome v0.6.4 ⇒ v0.6.1
  [3cdcf5f2] ↓ RecipesBase v1.0.2 ⇒ v0.8.0
  [348f2d5d] ↓ SpatialEcology v0.9.5 ⇒ v0.7.0
  [2913bbd2] ↓ StatsBase v0.33.0 ⇒ v0.32.2
```

### PR DESCRIPTION
**Pull request by carlobaldassi**
This is an improvement of the resolver heuristic, intended to fix #1949.

It is a strict improvement over the previous code, in terms of solution quality. It may be a tad slower, although I don't think it should be noticeable.
It fixed the problems mentioned in that issue that I could reproduce.
The PR is still not complete (there's still commented-out debugging stuff...), but the code is functional. So I'm posting here to encourage testing this (and reviewing, of course).

Still todo: polishing, tests, improve logging (?)

### CODE REVIEW

**Comment by KristofferC** [comment]
> We could compress it but that would introduce some testing dependencies, right?

Pkg has the ability to decompress so there is no extra test dependency. I pushed a commit that does this.

> The test is also kind of slow (a few minutes). Is it too much? Should I try to reduce it as much as I can? (That may take a while...)

We've worked quite a bit to reduce the test time, a full run takes about 8 minutes now. So a few minutes extra is a quite large percentage increase. But if it is a lot of work to reduce that test time I rather have this in now and get the improved test time later.

**Comment by carlobaldassi** [comment]
So I have some good news. I found out that my previous attempt was still suboptimal, in that it was missing the fact that in some cases we could have higher versions by installing additional packages. I have now further improved the optimizer to deal with this situation.
While doing that, I found out about a bug which went surprisingly unnoticed so far. Under very particular circumstances (such as some of those I used for testing this new code) it would produce a subtly inconsistent state, and potential crashes.
I have also reduced the test case as much as I could. It's now only a few kb and pretty fast. I checked that the new test would have failed with the previous optimizer.

So we should be all good hopefully, I have also cleaned up the code and added comments.

**Comment by KristofferC** [comment]
> in that it was missing the fact that in some cases we could have higher versions by installing additional packages

Yes, I've noticed that when this tend to happen, usually additional packages are installed.

**Comment by KristofferC** [comment]
I added a commit with a few fixes to things `@code_warntype` complained about.

File: src/Resolve/Resolve.jl [hunk 1]
```diff
@@ -411,11 +372,182 @@ function enforce_optimality!(sol::Vector{Int}, graph::Graph)
         staged = staged_next
     end
 
+    # @info "clearing up: $([pkgID(p0, graph) for p0 in findall(uninst) if sol[p0] ≠ spp[p0]])"
```
Comment (KristofferC) [inline src/Resolve/Resolve.jl:375]: Is this one (and the comment below) intended to be left in (to ease future debugging perhaps)?

Comment (carlobaldassi) [reply src/Resolve/Resolve.jl:375]: Ah no sorry, I wasn't intending to leave this one here. I'll clean it up.

**Comment by KristofferC** [comment]
Is it ok to squash this or do you prefer to keep the commits? If no squash, perhaps some of the commits should be merged together?

**Comment by carlobaldassi** [comment]
Squash away

**Pull request merged**